client/lib/tree-select/index.js

```diff
@@ -0,0 +1,72 @@
+/** @format */
+
+/**
+ * External dependencies
+ */
+import { isObject, some, isFunction } from 'lodash';
+
+/**
+ * Returns a selector that caches values.
+ *
+ * @param  {Function} getDependents A Function describing the dependent(s) of the selector.
+ *                                    Must return an array which gets passed as the first arg to the selector
+ * @param  {Function} selector      A standard selector for calculating cached result
+ * @return {Function}               Cached selector
+ */
+export default function treeSelect( getDependents, selector ) {
+	if ( ! isFunction( getDependents ) || ! isFunction( selector ) ) {
+		throw new TypeError(
+			'treeSelect: invalid arguments passed, selector and getDependents must both be functions'
+		);
+	}
+
+	const cache = new WeakMap();
+
+	const cachedSelector = function( state, ...args ) {
+		const dependents = getDependents( state, ...args );
+
+		if ( process.env.NODE_ENV !== 'production' ) {
+			if ( some( args, isObject ) ) {
+				throw new Error( 'Do not pass objects as arguments to a treeSelector' );
+			}
+		}
+
+		// create a dependency tree for caching selector results.
+		// this is beneficial over standard memoization techniques so that we can
+		// garbage collect any values that are based on outdated dependents
+		const cursor = dependents.reduce( insertDependentKey, cache );
+
+		const key = args.join();
+		if ( ! cursor.has( key ) ) {
+			cursor.set( key, selector( dependents, ...args ) );
+		}
+
+		return cursor.get( key );
```

Comment:
Here you already know the return value (it's the one that you just inserted into the map), so you can avoid the extra map lookup here.

Selectors are called zillion times on every possible Redux dispatch, so even a tiny optimization like this is very likely to be detectable.

Reply:
it'll require an extra line to assign the value to a variable.  I'll do that

Reply:
One more line for Jake, one less function call for the CPU. Looks like a good tradeoff 😜

Reply:
maybe. we probably shouldn't make too many assumptions about this though due to JS engine optimization. probably less CPU work but not definitely. maybe not significant but probably not.

Reply:
@dmsnell I'm pretty sure that the call to `Map.prototype.get` would be actually executed by the engine and that it takes more than zero CPU cycles ;)